commit d81b00e9a0f67ac992df1034d920477b5f8b7c1c
Author: Andrea Iacovitti <aiacovitti@libero.it>
Date:   Mon Mar 4 17:07:52 2013 +0100

    Take into account the clearButton width, that already takes some room of
    lineedit's contents, when rendering lineedit's padding.
    
    BUG: 316117
    FIXED-IN: 4.10.2

diff --git a/khtml/rendering/render_form.cpp b/khtml/rendering/render_form.cpp
index 2894b84..a5fce8e 100644
--- a/khtml/rendering/render_form.cpp
+++ b/khtml/rendering/render_form.cpp
@@ -91,6 +91,7 @@ using namespace DOM;
         {
             noBorder = false;
             left = right = top = bottom = 0;
+            clearButtonOverlay = 0;
             setParent(parent);
         }
 
@@ -101,7 +102,7 @@ using namespace DOM;
               case QStyle::SE_PushButtonContents:
               case QStyle::SE_LineEditContents:
               case QStyle::SE_ShapedFrameContents:
-                r.adjust(left, top, -right, -bottom);
+                r.adjust(left, top, -qMax(0, right - clearButtonOverlay), -bottom);
               default:
                 break;
             }
@@ -224,6 +225,7 @@ using namespace DOM;
         }
 
         int left, right, top, bottom;
+        int clearButtonOverlay;
         bool noBorder;
     };
 
@@ -1122,6 +1124,10 @@ void RenderLineEdit::setStyle(RenderStyle* _style)
             }
         }
     }
+
+    if (m_proxyStyle) {
+        static_cast<KHTMLProxyStyle*>(m_proxyStyle)->clearButtonOverlay = qMax(0, widget()->clearButtonUsedSize().width());
+    }
 }
 
 void RenderLineEdit::highLightWord( unsigned int length, unsigned int pos )
